src/stream/src/executor/arrangement_backfill.rs

```diff
@@ -0,0 +1,679 @@
+                yield msg;
+            }
+        }
+
+        // After progress finished + state persisted,
+        // we can forward messages directly to the downstream,
+        // as backfill is finished.
+        #[for_await]
+        for msg in upstream {
+            if let Some(msg) = Self::mapping_message(msg?, &self.output_indices) {
+                if let Some(state_table) = self.state_table.as_mut() && let Message::Barrier(barrier) = &msg {
+                        state_table.commit_no_data_expected(barrier.epoch);
+                    }
+                yield msg;
+            }
+        }
+    }
+
+    #[try_stream(ok = Option<StreamChunk>, error = StreamExecutorError)]
+    async fn snapshot_read(
+        schema: Arc<Schema>,
+        upstream_table: &StateTable<S>,
+        current_pos: Option<OwnedRow>,
+    ) {
+        // `current_pos` is None means it needs to scan from the beginning, so we use Unbounded to
+        // scan. Otherwise, use Excluded.
+        let range_bounds: (Bound<OwnedRow>, Bound<OwnedRow>) =
+            if let Some(current_pos) = current_pos {
+                // If `current_pos` is an empty row which means upstream mv contains only one row
+                // and it has been consumed. The iter interface doesn't support
+                // `Excluded(empty_row)` range bound, so we can simply return `None`.
+                if current_pos.is_empty() {
+                    assert!(upstream_table.pk_indices().is_empty());
+                    yield None;
+                    return Ok(());
+                }
+
+                (Bound::Excluded(current_pos), Bound::Unbounded)
+            } else {
+                (Bound::Unbounded, Bound::Unbounded)
+            };
+        let iterators = upstream_table
+            .iter_all_vnodes_with_pk_range(&range_bounds, Default::default())
+            .await?;
+        let pinned_iter: Vec<_> = iterators.into_iter().map(Box::pin).collect_vec();
+        let iter = merge_sort(pinned_iter);
+        pin_mut!(iter);
+        while let Some(data_chunk) = collect_data_chunk(&mut iter, &schema, Some(CHUNK_SIZE))
+            .instrument_await("arrangement_backfill_snapshot_read")
+            .await?
+        {
+            if data_chunk.cardinality() != 0 {
+                let ops = vec![Op::Insert; data_chunk.capacity()];
+                let stream_chunk = StreamChunk::from_parts(ops, data_chunk);
+                yield Some(stream_chunk);
+            }
+        }
+
+        yield None;
+    }
+
+    /// Mark chunk:
+    /// For each row of the chunk, forward it to downstream if its pk <= `current_pos`, otherwise
+    /// ignore it. We implement it by changing the visibility bitmap.
+    fn mark_chunk(
+        chunk: &StreamChunk,
+        current_pos: &OwnedRow,
+        pk_in_output_indices: PkIndicesRef<'_>,
+        pk_order: &[OrderType],
+    ) -> StreamChunk {
+        // FIXME: Temporary workaround, we can optimize it later when benchmarking.
```

Comment:
Take note about this, we no longer have ownership of stream chunk here, because after yielding, we still need to write + flush the chunk to our local state store.

I prefer to write + flush chunk after yielding, so that write + flush can have ownership of the chunk.

Here optimizing `chunk.compact()` to `chunk.compact_ref()` will mean minimal performance impact, in comparison to optimizing it when write + flush chunk.